src/bidiMapper/utils/log.ts

```diff
@@ -16,17 +16,27 @@
 export function log(type: string): (...message: any[]) => void {
   return (...messages: any[]) => {
     // If run in browser, add debug message to the page.
-    if (typeof window !== 'undefined' && window?.document?.documentElement) {
+    if (globalThis?.document?.documentElement) {
       console.log(type, ...messages);
 
-      const elementId = type + '_log';
+      const typeLogContainer = findOrCreateTypeLogContainer(type);
 
-      if (!window.document.getElementById(elementId)) {
-        window.document.documentElement.innerHTML += `<h3>${type}:</h3><pre id='${elementId}'></pre>`;
-      }
-
-      const element = window.document.getElementById(elementId);
-      element.innerText += messages.join(', ') + '\n';
+      const pre = document.createElement('pre');
+      pre.innerText = messages.join(', ');
+      typeLogContainer.appendChild(pre);
     }
   };
 }
+
+function findOrCreateTypeLogContainer(type: string) {
+  const elementId = type + '_log';
+  if (!document.getElementById(elementId)) {
+    const typeLogContainer = document.createElement('div');
+    typeLogContainer.id = elementId;
+    typeLogContainer.innerHTML = `<h3>${type}:</h3>`;
+    document.body.appendChild(typeLogContainer);
+    return typeLogContainer;
+  }
+
+  return document.getElementById(elementId);
```

Comment:
Optional nit: do just a single `document.getElementById(elementId)` call and return it if true:
```js
const something = document.getElementById(elementId);
if (something) {
  return something;
}
```

Reply:
done